### ISSUE TITLE
[release/2.11] timm_regnet fail_accuracy

### ISSUE THREAD

**Issue by bjarzemb** [intel/torch-xpu-ops#2908]
### 🐛 Describe the bug

| Testcase | Model | Status | PT2.10 | PT2.11 | PT2.11 & old oneAPI | Ref log | Cur log |
|----------|-------|--------|--------|--------|---------------------|---------|---------|
| huggingface_accuracy_inference_amp-bf16 | DistillGPT2 | FAIL | pass | fail_accuracy | fail_accuracy | [log](https://github.com/user-attachments/files/25579767/e2e_huggingface_accuracy_inference_amp-bf16_DistillGPT2.log) | [log](https://github.com/user-attachments/files/25579953/e2e_huggingface_accuracy_inference_amp-bf16_DistillGPT2.log) |
| huggingface_accuracy_inference_amp-bf16 | AllenaiLongformerBase | FAIL | pass | fail_accuracy | fail_accuracy | [log](https://github.com/user-attachments/files/25579879/e2e_huggingface_accuracy_inference_amp-bf16_AllenaiLongformerBase.log) | [log](https://github.com/user-attachments/files/25579970/e2e_huggingface_accuracy_inference_amp-bf16_AllenaiLongformerBase.log) |
| huggingface_accuracy_inference_amp-bf16 | openai/whisper-tiny | FAIL | pass | fail_accuracy | fail_accuracy | [log](https://github.com/user-attachments/files/25579890/e2e_huggingface_accuracy_inference_amp-bf16_openai.whisper-tiny.log) | [log](https://github.com/user-attachments/files/25579973/e2e_huggingface_accuracy_inference_amp-bf16_openai.whisper-tiny.log) |
| timm_models_accuracy_inference_amp-fp16 | pit_b_224 | FAIL | pass | fail_accuracy | fail_accuracy | [log](https://github.com/user-attachments/files/25579886/e2e_timm_models_accuracy_inference_amp-fp16_pit_b_224.log) | [log](https://github.com/user-attachments/files/25579979/e2e_timm_models_accuracy_inference_amp-fp16_pit_b_224.log) |
| torchbench_accuracy_training_bfloat16 | timm_regnet | FAIL | pass | fail_accuracy | fail_accuracy | [log](https://github.com/user-attachments/files/25579888/e2e_torchbench_accuracy_training_bfloat16_timm_regnet.log) | [log](https://github.com/user-attachments/files/25579982/e2e_torchbench_accuracy_training_bfloat16_timm_regnet.log) |

### Versions

PyTorch version: 2.11.0+xpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro (10.0.26100 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: N/A

Python version: 3.10.19 | packaged by conda-forge | (main, Jan 26 2026, 23:39:36) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.26100-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: True
XPU used to build PyTorch: 20250302
Intel GPU driver version:
* 32.0.101.8331 (20251126000000.******+***)
Intel GPU models onboard:
* Intel(R) Arc(TM) B580 Graphics
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Arc(TM) B580 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE20B, uuid=86800be2-0000-0000-0400-000000000000, driver_version='1.13.35716', total_memory=11869MB, local_mem_size=128KB, max_compute_units=160, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Name: Intel(R) Core(TM) Ultra 5 245K
Manufacturer: GenuineIntel
Family: 773
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 3071
MaxClockSpeed: 4200
L2CacheSize: 26624
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] dpcpp-cpp-rt==2025.3.2
[pip3] intel-cmplr-lib-rt==2025.3.2
[pip3] intel-cmplr-lib-ur==2025.3.2
[pip3] intel-cmplr-lic-rt==2025.3.2
[pip3] intel-opencl-rt==2025.3.2
[pip3] intel-openmp==2025.3.2
[pip3] intel-pti==0.16.0
[pip3] intel-sycl-rt==2025.3.2
[pip3] mkl==2025.3.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.23.2
[pip3] onemkl-license==2025.3.1
[pip3] onemkl-sycl-blas==2025.3.1
[pip3] onemkl-sycl-dft==2025.3.1
[pip3] onemkl-sycl-lapack==2025.3.1
[pip3] onemkl-sycl-rng==2025.3.1
[pip3] onemkl-sycl-sparse==2025.3.1
[pip3] onnx==1.20.0
[pip3] onnx-ir==0.1.16
[pip3] onnxscript==0.6.2
[pip3] optree==0.13.0
[pip3] tbb==2022.3.1
[pip3] tcmlib==1.4.1
[pip3] torch==2.11.0+xpu
[pip3] torchaudio==2.11.0+xpu
[pip3] torchvision==0.26.0+xpu
[pip3] triton-xpu==3.7.0
[pip3] umf==1.0.3
[conda] dpcpp-cpp-rt              2025.3.2                 pypi_0    pypi
[conda] intel-cmplr-lib-rt        2025.3.2                 pypi_0    pypi
[conda] intel-cmplr-lib-ur        2025.3.2                 pypi_0    pypi
[conda] intel-cmplr-lic-rt        2025.3.2                 pypi_0    pypi
[conda] intel-opencl-rt           2025.3.2                 pypi_0    pypi
[conda] intel-openmp              2025.3.2                 pypi_0    pypi
[conda] intel-pti                 0.16.0                   pypi_0    pypi
[conda] intel-sycl-rt             2025.3.2                 pypi_0    pypi
[conda] mkl                       2025.3.1                 pypi_0    pypi
[conda] numpy                     1.23.2                   pypi_0    pypi
[conda] onemkl-license            2025.3.1                 pypi_0    pypi
[conda] onemkl-sycl-blas          2025.3.1                 pypi_0    pypi
[conda] onemkl-sycl-dft           2025.3.1                 pypi_0    pypi
[conda] onemkl-sycl-lapack        2025.3.1                 pypi_0    pypi
[conda] onemkl-sycl-rng           2025.3.1                 pypi_0    pypi
[conda] onemkl-sycl-sparse        2025.3.1                 pypi_0    pypi
[conda] optree                    0.13.0                   pypi_0    pypi
[conda] tbb                       2022.3.1                 pypi_0    pypi
[conda] tcmlib                    1.4.1                    pypi_0    pypi
[conda] torch                     2.11.0+xpu               pypi_0    pypi
[conda] torchaudio                2.11.0+xpu               pypi_0    pypi
[conda] torchvision               0.26.0+xpu               pypi_0    pypi
[conda] triton-xpu                3.7.0                    pypi_0    pypi
[conda] umf                       1.0.3                    pypi_0    pypi

**Comment by mengfei25** [comment]
pit_b_224 also got fail_accuracy on BMG Linux, others are passed

**Comment by eryk-roch** [comment]
On Windows, XPU Triton kernels are compiled to SPIR-V , while on Linux they use ZE Binary. These different compiler backends can produce code with different floating-point rounding behavior for bfloat16 accumulations. This would explain why timm_regnet fails only on Windows.

@bjarzemb Can you verify by running the failing test with zebin format on Windows?

set TORCHINDUCTOR_XPU_KERNEL_FORMAT=zebin

**Comment by bjarzemb** [comment]
I verified failing test cases on windows with following versions:
* PyTorch commit: 9ed8fd28 (from 2026-05-15)
* driver: 32.0.101.8801
* DLE: 2025.3.2.61

without setting TORCHINDUCTOR_XPU_KERNEL_FORMAT:
Testcase | Model | Result
-- | -- | --
timm_models_accuracy_inference_amp-fp16 | pit_b_224 | fail_accuracy
torchbench_accuracy_training_bfloat16 | timm_regnet | pass


I removed triton cache and run pit_b_224 with TORCHINDUCTOR_XPU_KERNEL_FORMAT=zebin, but test status is also fail_accuracy. Should I rebuild PyTorch wheel?

**Comment by eryk-roch** [comment]
No need to rebuild the PyTorch wheel,  TORCHINDUCTOR_XPU_KERNEL_FORMAT is a runtime environment variable, the build doesn't matter. Since pit_b_224 fails with both spv and zebin, the kernel format isn't the cause.

**Comment by eryk-roch** [comment]
Good news: timm_regnet passing confirms that issue is resolved with the newer PyTorch/driver. We can close that part.

**Comment by eryk-roch** [comment]
E0227 15:26:13.175000 9912 site-packages\torch_dynamo\utils.py:3393] RMSE (res-fp64): 0.11101, (ref-fp64): 0.00047 and shape=torch.Size([8, 1000]). res.dtype: torch.float16, multiplier: 3.000000, tol: 0.001000, use_larger_multiplier_for_smaller_tensor: 0 --> fail_accuracy

It looks like completely different results that has nothing to do with each other.

**Comment by eryk-roch** [comment]
It uses attention over 962 tokens (aten._softmax with shape [B, 4, 962, 962]), which goes through inductor's online_softmax Triton kernel (enabled by default for XPU in torch/_inductor/fx_passes/post_grad.py).

**Note:** the native ATen softmax on XPU already has known accuracy issues. Multiple test_compare_cpu_softmax_* tests are skipped in test/xpu/extended/skip_list_win_mtl.py. The Triton online_softmax is a different code path but may share underlying precision problems on this setup.|

@bjarzemb  Can you re-test it with softmax disabled? (set TORCHINDUCTOR_ONLINE_SOFTMAX=0)
If it passes, the root cause is the online_softmax Triton kernel on XPU with the 962-element reduction.
If it still fails, the issue is elsewhere in the compiled graph. Try also with set TORCHINDUCTOR_MEMORY_PLANNING=0 to rule out buffer reuse corruption.

**Comment by bjarzemb** [comment]
@eryk-roch I have also fail for TORCHINDUCTOR_ONLINE_SOFTMAX=0 and for TORCHINDUCTOR_MEMORY_PLANNING=0

**Comment by eryk-roch** [comment]
@bjarzemb thank you for retest and sharing logs. I can see there that pit_b_224 has complex fused Triton kernels that combine reading channels-last depthwise conv output with layer_norm + cat + slice + transpose. This fusion is the prime suspect.

Can you re-test adding this flag to the benchmark command:
-c max_fusion_size=1
?

**Comment by bjarzemb** [comment]
@eryk-roch I have pass with "-c max_fusion_size=1".

- PyTorch commit: 5ef6fae9 (from 2026-06-09)
- driver: 32.0.101.8801
- DLE: 2025.3.2.61

**Comment by anmyachev** [comment]
> timm_regnet is Windows only failure. pit_b_224 random failed on linux

**FYI** for pit_b_224 (on linux) we've narrowed it down to https://jira.devtools.intel.com/browse/GSD-12872

**Comment by eryk-roch** [comment]
This defect is covered by the IGC bug filed on WSL/Linux:
[ [TritonXPU] IGC optimization produces nondeterministic results for a LayerNorm kernel on PVC ](https://jira.devtools.intel.com/browse/GSD-12872)
Same root cause.  IGC optimizer  the fused LayerNorm reduction kernel generated by Triton for pit_b_224. Windows and WSL/Linux share the same IGC backend (libigc), so this will be fixed when that Jira ticket is resolved. No separate Windows-specific fix is needed.

**Comment by eryk-roch** [comment]
This defect is blocked by: https://github.com/intel/intel-xpu-backend-for-triton/issues/7043

Root cause: PyTorch Inductor generates a fused Triton kernel that stores data to a buffer and reads it back from a different warp without a barrier in between. The read can occur before the write is visible, a cross-subgroup data race on global memory.

Fix: https://github.com/pytorch/pytorch/pull/190519

@bjarzemb, please retest pit_b_224 accuracy on Windows once PR #190519 is merged and picked into the next release.

Workaround (until fix lands): TRITON_INTEL_DISABLE_IGC_OPT=1 or -c max_fusion_size=1

**Comment by bjarzemb** [comment]
Unfortunately I still have timm_regnet fail_accuracy in Weekly tests.
PyTorch: 140e4d5 (Jul 9)
torch-xpu-ops: d87e686f (Jul 8)

Log:
```
loading model: 0it [00:00, ?it/s]
loading model: 0it [00:06, ?it/s]
xpu  train timm_regnet                        
W0710 16:29:50.487000 10060 site-packages\torch\_inductor\cudagraph_utils.py:401] [2/0_1] [__cudagraphs] skipping cudagraphs due to skipping cudagraphs due to multiple devices: device(type='xpu', index=0)
E0710 16:31:47.298000 10060 site-packages\torch\_dynamo\utils.py:3736] RMSE (res-fp64): 0.01380, (ref-fp64): 0.00030 and shape=torch.Size([]). res.dtype: torch.float32, multiplier: 3.000000, tol: 0.001000, use_larger_multiplier_for_smaller_tensor: 0
fail_accuracy
```

**Comment by bjarzemb** [comment]
pit_b_224 tracked in https://github.com/intel/torch-xpu-ops/issues/4350

**Comment by bjarzemb** [comment]
timm_regnet passed 5/5 times on:
PyTorch: fddcbe3 (Jul 28)
torch-xpu-ops: 298168d8 (Jul 24)
